From dfc243c394cb55424b7026cc86723819675c5e9b Mon Sep 17 00:00:00 2001
From: Hains van den Bosch <hainsvdbosch@ziggo.nl>
Date: Fri, 13 Mar 2026 17:49:42 +0100
Subject: [PATCH] Fix build against glibc 2.43

https://lists.gnu.org/archive/html/info-gnu/2026-01/msg00005.html

| locate_kdc.c: In function 'parse_uri_if_https':
| locate_kdc.c:226:12: error: assignment discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]
|   226 |         cp = strchr(*host, '/');
|       |            ^
---
 src/lib/krb5/ccache/ccbase.c                       | 4 ++--
 src/lib/krb5/os/expand_path.c                      | 4 ++--
 src/lib/krb5/os/locate_kdc.c                       | 7 ++++---
 src/plugins/preauth/pkinit/pkinit_crypto.h         | 2 +-
 src/plugins/preauth/pkinit/pkinit_crypto_openssl.c | 6 +++---
 src/plugins/preauth/pkinit/pkinit_identity.c       | 2 +-
 src/plugins/preauth/pkinit/pkinit_matching.c       | 2 +-
 7 files changed, 14 insertions(+), 13 deletions(-)

diff --git a/src/lib/krb5/ccache/ccbase.c b/src/lib/krb5/ccache/ccbase.c
index 696b681..9fc7d10 100644
--- a/src/lib/krb5/ccache/ccbase.c
+++ b/src/lib/krb5/ccache/ccbase.c
@@ -201,8 +201,8 @@ krb5_cc_register(krb5_context context, const krb5_cc_ops *ops,
 krb5_error_code KRB5_CALLCONV
 krb5_cc_resolve (krb5_context context, const char *name, krb5_ccache *cache)
 {
-    char *pfx, *cp;
-    const char *resid;
+    char *pfx;
+    const char *resid, *cp;
     unsigned int pfxlen;
     krb5_error_code err;
     const krb5_cc_ops *ops;
diff --git a/src/lib/krb5/os/expand_path.c b/src/lib/krb5/os/expand_path.c
index 5cbccf0..51bc266 100644
--- a/src/lib/krb5/os/expand_path.c
+++ b/src/lib/krb5/os/expand_path.c
@@ -454,8 +454,8 @@ k5_expand_path_tokens_extra(krb5_context context, const char *path_in,
 {
     krb5_error_code ret;
     struct k5buf buf;
-    char *tok_begin, *tok_end, *tok_val, **extra_tokens = NULL, *path;
-    const char *path_left;
+    char *tok_val, **extra_tokens = NULL, *path;
+    const char *path_left, *tok_begin, *tok_end;
     size_t nargs = 0, i;
     va_list ap;
 
diff --git a/src/lib/krb5/os/locate_kdc.c b/src/lib/krb5/os/locate_kdc.c
index d1df04a..fc9fc5d 100644
--- a/src/lib/krb5/os/locate_kdc.c
+++ b/src/lib/krb5/os/locate_kdc.c
@@ -217,7 +217,8 @@ static void
 parse_uri_if_https(const char *host_or_uri, k5_transport *transport,
                    const char **host, const char **uri_path)
 {
-    char *cp;
+    const char *cp;
+    char *cp2;
 
     if (strncmp(host_or_uri, "https://", 8) == 0) {
         *transport = HTTPS;
@@ -225,8 +226,8 @@ parse_uri_if_https(const char *host_or_uri, k5_transport *transport,
 
         cp = strchr(*host, '/');
         if (cp != NULL) {
-            *cp = '\0';
-            *uri_path = cp + 1;
+            *cp2 = '\0';
+            *uri_path = cp2 + 1;
         }
     }
 }
diff --git a/src/plugins/preauth/pkinit/pkinit_crypto.h b/src/plugins/preauth/pkinit/pkinit_crypto.h
index 57bb3cb..f6b6a59 100644
--- a/src/plugins/preauth/pkinit/pkinit_crypto.h
+++ b/src/plugins/preauth/pkinit/pkinit_crypto.h
@@ -440,7 +440,7 @@ krb5_error_code crypto_load_cas_and_crls
 		    defines the storage type (file, directory, etc) */
 	int catype,					/* IN
 		    defines the ca type (anchor, intermediate, crls) */
-	char *id);					/* IN
+	const char *id);					/* IN
 		    defines the location (filename, directory name, etc) */
 
 /*
diff --git a/src/plugins/preauth/pkinit/pkinit_crypto_openssl.c b/src/plugins/preauth/pkinit/pkinit_crypto_openssl.c
index f222dbd..0a731a8 100644
--- a/src/plugins/preauth/pkinit/pkinit_crypto_openssl.c
+++ b/src/plugins/preauth/pkinit/pkinit_crypto_openssl.c
@@ -4999,7 +4999,7 @@ load_cas_and_crls(krb5_context context,
                   pkinit_req_crypto_context req_cryptoctx,
                   pkinit_identity_crypto_context id_cryptoctx,
                   int catype,
-                  char *filename)
+                  const char *filename)
 {
     STACK_OF(X509_INFO) *sk = NULL;
     STACK_OF(X509) *ca_certs = NULL;
@@ -5157,7 +5157,7 @@ load_cas_and_crls_dir(krb5_context context,
                       pkinit_req_crypto_context req_cryptoctx,
                       pkinit_identity_crypto_context id_cryptoctx,
                       int catype,
-                      char *dirname)
+                      const char *dirname)
 {
     krb5_error_code retval = EINVAL;
     char **fnames = NULL, *filename;
@@ -5201,7 +5201,7 @@ crypto_load_cas_and_crls(krb5_context context,
                          pkinit_identity_crypto_context id_cryptoctx,
                          int idtype,
                          int catype,
-                         char *id)
+                         const char *id)
 {
     switch (idtype) {
     case IDTYPE_FILE:
diff --git a/src/plugins/preauth/pkinit/pkinit_identity.c b/src/plugins/preauth/pkinit/pkinit_identity.c
index 0dcfcfc..ad65f23 100644
--- a/src/plugins/preauth/pkinit/pkinit_identity.c
+++ b/src/plugins/preauth/pkinit/pkinit_identity.c
@@ -473,7 +473,7 @@ process_option_ca_crl(krb5_context context,
                       const char *value,
                       int catype)
 {
-    char *residual;
+    const char *residual;
     unsigned int typelen;
     int idtype;
 
diff --git a/src/plugins/preauth/pkinit/pkinit_matching.c b/src/plugins/preauth/pkinit/pkinit_matching.c
index 0ea072c..b3c8df1 100644
--- a/src/plugins/preauth/pkinit/pkinit_matching.c
+++ b/src/plugins/preauth/pkinit/pkinit_matching.c
@@ -262,7 +262,7 @@ parse_rule_component(krb5_context context,
     char err_buf[128];
     int ret;
     struct keyword_desc *kw, *nextkw;
-    char *nk;
+    const char *nk;
     int found_next_kw = 0;
     char *value = NULL;
     size_t len;
-- 
2.51.0

